### PR TITLE
GL-919: Implement throttling GL api

### PR DESCRIPTION
**Pull request by rasikasri**
### Jira link

[GL-919](https://transformuk.atlassian.net/browse/GL-919)

### What?

I have added/removed/altered:

- [ ] Implemented throttleing API request base on api key

### Why?

I am doing this because:

- API need to be protected from heavy load, hence it needs a throtteling mechanism

### CODE REVIEW

**Review by neilmiddleton — COMMENTED** [review]
Are there tests coming?

File: README.md [hunk 1]
```diff
@@ -84,6 +84,31 @@ TARIFF_SYNC_USERNAME
 GREEN_LANES_UPDATE_EMAIL
 ```
 
+### API Key Management
```
Comment (willfish) [inline README.md:87]: I don't think this will pass the markdown linting rules?

Comment (willfish) [reply README.md:87]: How do you feel about putting a separate doc under docs/green_lanes_api_keys.md or equivalent?

File: config/initializers/rack_attack.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    end
+
+    throttle('default', limit: 10, period: 1.hour) do |req|
+      # Apply to all requests with an API key not in the list
+      api_key = req.env['HTTP_X_API_KEY']
+      API_KEY_LIMITS.blank? || (!API_KEY_LIMITS.key?(api_key) && api_key.present?)
+    end
+  end
+else
+  logger.info 'Rack::Attack is disabled in Dev env.'
```
Comment (willfish) [inline config/initializers/rack_attack.rb:23]: I didn't know `logger` was injected into the scope of an initialiser

File: config/initializers/rack_attack.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+if Rails.env.production?
+  class Rack::Attack
+    Rack::Attack.cache.store = ActiveSupport::Cache::RedisCacheStore.new(url: ENV['REDIS_URL'])
+
+    API_KEY_LIMITS = JSON.parse(ENV['GREEN_LANES_API_KEYS'])
```
Comment (willfish) [inline config/initializers/rack_attack.rb:5]: I'm a little lost on how this controls just for /green_lanes/ scoped api calls?

File: config/initializers/rack_attack.rb (hunk 2 above)
Comment (rasikasri) [reply config/initializers/rack_attack.rb:16]: I double checked and logger is available there

File: config/initializers/rack_attack.rb (hunk 3 above)
Comment (rasikasri) [reply config/initializers/rack_attack.rb:5]: This is a good point, I added logic to only throttle `/green_lanes/goods_nomenclatures` endpoint. Also removed the default throttle method as it doesn't make sense. If the api-key does not exist in the json, the the request will be blocked anyway.

**Review by willfish — APPROVED** [review]
It would be good to do a performance benchmark with/without this enabled (e.g. on staging) before/after we merge this

**Pull request merged**